Use `@isolated(any)` function types in task creation

rjmccall opened the pull request.

rjmccall committed:
Properly save and restore the current task in the runtime so that tasks
can be reentrantly executed.

I don't think doing this is *actually a good idea*, but corrupting the
runtime is an even worse idea, and the overhead here is very low.

rjmccall committed:
Use @isolated(any) function types for task functions.

The biggest annoyance here is having to clone all of the task creation
functions for Embedded Swift because it can't use `any Actor` right now.

rjmccall commented:
@swift-ci Please test

ktoso commented on stdlib/public/Concurrency/DiscardingTaskGroup.swift:
```diff
@@ -604,9 +827,10 @@ public struct ThrowingDiscardingTaskGroup<Failure: Error> {
   public mutating func addTask(
     priority: TaskPriority? = nil,
-    operation: __owned @Sendable @escaping () async throws -> Void
+    operation: __owned @Sendable @escaping @isolated(any) () async throws -> Void
```
maybe just not implemented yet (?), would we want to remove the `@Sendable` since isolated(any) will imply Sendable?

rjmccall replied:
It's not implemented yet, but also the sanest approach for the feature-suppression mechanism is to just decline to print `@isolated(any)` and not to try to add back `@Sendable`, so we do need to keep the redundant `@Sendable` in there.

ktoso replied:
Gotcha, makes sense!

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -299,6 +309,81 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
+#else
+    let flags = taskCreateFlags(
+      priority: priority, isChildTask: true, copyTaskLocals: false,
+      inheritContext: false, enqueueJob: true,
```
Just an idea: I wonder if it's time to make some

```swift
@inline(__always)
@inlinable
var SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY: Bool { 
#if SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
  true
#else 
  false
#endif 
}
```

because this entire #ifelse is really just to change the `enqueueJob` depending on the env...?

(Trying to get less of those #ifelse in this file, it's getting somewhat intense...)

rjmccall replied:
Yeah, think there's a lot of redundancy we could eliminate with the flags.

ktoso replied:
Yeap some soon™ cleanups then... No problem in this PR though, land as is :)

ktoso reviewed (approved):
Always rough to follow all the #ifs in there but it looks good 👍

al45tair commented on include/swift/Threading/ThreadLocalStorage.h:
```diff
@@ -115,6 +115,15 @@ class ThreadLocal {
 
   void set(T newValue) { value = newValue; }
+
+  T replace(T newValue) {
```
Minor: would `swap` be a better name for this, given that we return the old value? (I'm not sure, just pondering whether there's a way to somehow make it obvious that it returns the original value.)

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Pass the task option record pointer to swift_task_create as a pointer.

We add the `memory(argmem: readwrite)` attribute to swift_task_create,
which means that the call is only allowed to read or write "pointer
operands".  LLVM is smart enough to look through obvious ptrtoint
casts, but not to look through integer selects and so on, which is what
we produce when there's an opaque optional operand that feeds into the
builtin.  This was causing miscompiles under optimization when using
`@isolated(any)` function types for task creation, since we're not yet
clever enough to fold the function_extract_isolation for a known function
(and of course it's not necessarily a known function anyway).

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Make `@isolated(any)` imply `@Sendable`. No avoidance mechanism for now.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

hjyamauchi commented:
It seems that this issue https://github.com/apple/swift/issues/72484 was caused by this commit. Any thoughts?

hjyamauchi commented:
Here's a reduced test: https://github.com/apple/swift/issues/72484#issuecomment-2019047017
